### ISSUE TITLE
Improve keyboard navigation

### ISSUE THREAD

**Issue by wojciechczerniak** [handsontable/handsontable#8729]
### Description
We want to improve overall keyboard navigation.

### Goals

- [ ] Make it possible to navigate headers with keyboard (#8939)
- [ ] A ShortcutManager that is able to switch between context: grid, pagination, renderer, editor and provide the right combination of shortcuts for them (#8942)

### Component focus management

- [ ] <kbd>Tab</kbd> and <kbd>Shift</kbd>+<kbd>Tab</kbd> should manage component focus only
- [ ] When HOT is not focused as a component, it's a part of web page / web application tab flow 
- [ ] When focus is set to an element prior our component, <kbd>Tab</kbd> will focus the component and subsequent <kbd>Tab</kbd> key will navigate out of the component to an element that is after it
- [ ] When focus is set to an element after our component, <kbd>Shift</kbd>+<kbd>Tab</kbd> will focus the component and subsequent <kbd>Shift</kbd>+<kbd>Tab</kbd> key combination will navigate out of the component to an element that is before it
- [ ] While changing the focus out/in the component it is important to focus on a last active element that user was on. This requirement does not change whether `outsideClickDeselects` is active or not.
- [ ] If it's the first time the component is being focused on, most top and left cell should be selected by default
- [ ] All elements that normally would be included in web page / web app tab flow, all active elements like `button`, `a`, and others, should have `tabindex="-1"` set so they are excluded from it
- [x] ~Handsontable does not have header, footre, statusbar, pagination, and/or sidebar panels that are considered as a separate widget in <kbd>Tab</kbd> key flow, so above simple rules should be sufficient.~

### Grid navigation

Once component is focused we're in grid navigation mode. Data cells can be focused using standard navigation keys, including directional arrows, Home, End, Control + Home, and Control + End. Page Up and Page Down scroll the grid

- [ ] <kbd>Right Arrow</kbd>: Moves focus one cell to the right. If focus is on the right-most cell in the row, focus does not move.
- [ ] <kbd>Left Arrow</kbd>: Moves focus one cell to the left. If focus is on the left-most cell in the row, focus does not move.
- [ ] <kbd>Down Arrow</kbd>: Moves focus one cell down. If focus is on the bottom cell in the column, focus does not move.
- [ ] <kbd>Up Arrow</kbd>: Moves focus one cell Up. If focus is on the top cell in the column, focus does not move.
- [ ] <kbd>Page Down</kbd>: Moves focus down so the bottom row in the currently visible set of rows becomes one of the first visible rows. If focus is in the last row of the grid, focus does not move.
- [ ] <kbd>Page Up</kbd>: Moves focus up so the top row in the currently visible set of rows becomes one of the last visible rows. If focus is in the first row of the grid, focus does not move.
- [ ] <kbd>Home</kbd>: Moves focus to the first cell in the row that contains focus.
- [ ] <kbd>End</kbd>: Moves focus to the last cell in the row that contains focus.
- [ ] <kbd>Control</kbd>+<kbd>Home</kbd>: Moves focus to the first cell in the first row.
- [ ] <kbd>Control</kbd>+<kbd>End</kbd>: Moves focus to the last cell in the last row.
- [ ] Context Menu should open on <kbd>Fn</kbd>+<kbd>Control</kbd>+<kbd>i</kbd> on Mac, <kbd>Shift</kbd>+<kbd>F10</kbd> on Windows, and dedicated <kbd>Menu</kbd> key. Context Menu is activated for focused cell or header
- [ ] <kbd>Enter</kbd> or <kbd>F2</kbd> activates the cell. More below

### Cell navigation

- [ ] Each cell content is locked until activated, either by Enter or F2 key
- [ ] Alphanumeric keys: may activate the editor if there is an input field that can be focused
- [ ] <kbd>Enter</kbd>: If the cell contains editable content, places focus in an input field, such as a textbox. If the input is a single-line text field, a subsequent press of Enter may either restore grid navigation functions or move focus to an input field in a neighboring cell.
- [ ] <kbd>Enter</kbd>: If the cell contains one or more widgets, places focus on the first widget.
- [ ] <kbd>F2</kbd>: If a cell contains editable content, place focus in an input field. A subsequent press of F2 restores grid navigation functions.
- [ ] <kbd>F2</kbd>: If a cell contains one or more widgets, place focus on the first widget. A subsequent press of F2 restores grid navigation functions.
- [ ] When grid navigation is disabled <kbd>Escape</kbd> key: restore grid navigation. If content was being edited, it may also undo/cancel the changes.

### Header navigation

- [ ] Each cell content is locked until activated, either by Enter or F2 key
- [ ] Alphanumeric keys: may activate the editor if there is an input field that can be focused
- [ ] <kbd>Enter</kbd>: If the cell contains editable content, places focus in an input field, such as a textbox. If the input is a single-line text field, a subsequent press of Enter may either restore grid navigation functions or move focus to an input field in a neighboring cell.
- [ ] <kbd>Enter</kbd>: If the cell contains one or more widgets, places focus on the first widget.
- [ ] <kbd>F2</kbd>: If a cell contains editable content, place focus in an input field. A subsequent press of F2 restores grid navigation functions.
- [ ] <kbd>F2</kbd>: If a cell contains one or more widgets, place focus on the first widget. A subsequent press of F2 restores grid navigation functions.
- [ ] When grid navigation is disabled <kbd>Escape</kbd> key: restore grid navigation. If content was being edited, it may also undo/cancel the changes.
- [ ] Header cells should have focus style that should be consistent with cell focus. At the same time, cell should have the focus style removed / moved to header cell and look as it is less active.
- [ ] Moving navigation focus to header should not reject current selection

### Selection

- [ ] <kbd>Control</kbd> + <kbd>Space</kbd>: selects the column that contains the focused cell / header.
- [ ] <kbd>Shift</kbd> + <kbd>Space</kbd>: Selects the row that contains the focused cell / header. If the grid includes a column with checkboxes for selecting rows, this key can serve as a shortcut for checking the box when focus is not on the checkbox.
- [ ] <kbd>Control</kbd> + <kbd>A</kbd>: Selects all cells.
- [ ] <kbd>Shift</kbd> + <kbd>Right Arrow</kbd>: Extends selection one cell to the right.
- [ ] <kbd>Shift</kbd> + <kbd>Left Arrow</kbd>: Extends selection one cell to the left.
- [ ] <kbd>Shift</kbd> + <kbd>Down Arrow</kbd>: Extends selection one cell down.
- [ ] <kbd>Shift</kbd> + <kbd>Up Arrow</kbd>: Extends selection one cell Up.
- [ ] Selection have to be visually distinct from focus style

### Autofill

- [ ] <kbd>Shift</kbd> + <kbd>Right Arrow</kbd> to expand selection, then <kbd>Ctrl</kbd> + <kbd>R</kbd> to fill right, first active cell only
- [ ] <kbd>Shift</kbd> + <kbd>Down Arrow</kbd> to expand selection, then <kbd>Ctrl</kbd> + <kbd>D</kbd> to fill down, first active cell only
- [ ] <kbd>Shift</kbd> + <kbd>Space</kbd> to select row, then <kbd>Ctrl</kbd> + <kbd>R</kbd> to fill right, first cell only
- [ ] <kbd>Control</kbd> + <kbd>Space</kbd> to select column, then <kbd>Ctrl</kbd> + <kbd>D</kbd> to fill down, first cell only

### Clipboard

- [x] <kbd>Control</kbd> + <kbd>C</kbd>, <kbd>Control</kbd> + <kbd>X</kbd>, and <kbd>Control</kbd> + <kbd>V</kbd> work as they should
- [ ] <kbd>Control</kbd> + <kbd>Shift</kbd> + <kbd>V</kbd> should paste values only

### Undo / Redo

- [x] <kbd>Control</kbd> + <kbd>Z</kbd> works as it should

### Features not addressed here yet

- ~Non-contiguous selection. We need a keyboard shortcut to add selection layer~ addressed in handsontable/dev-handsontable#383
- Manual(Row / Column)Move. Header will require drag handle or dropdown menu entry
- Manual(Row / Column)Resize. A context menu option could help
- Navigation inside editors and renderers. This spec is purely for grid layout
- No idea how to display comment pop-up
- Resize may be done by modal

**Comment by krzysztofspilka** [comment]
>No idea how to display comment pop-up

@wojciechczerniak One idea is to make it similar to spreadsheets (E), e.g.:

TASK | ALT SEQUENCE
-- | --
New Comment | Alt-R-C
Edit Comment | Alt-R-T
Delete Comment | Alt-R-D
Show/Hide Comment | Alt-R-H*
Show All Comments | Alt-R-A
